File: modules/Engine/src/Engine/GameState/Game.cpp
#include <Engine/GameState/GameState.hpp>
#include <Engine/Resources.hpp>
#include <Engine/Settings.hpp>

#include <Util/Observer.hpp>

#include <iostream>

namespace engine::GameState
{

Game::Game()
: currentLevel(dungeonGen)
{
        this->currentLevel.setPosition({8.f, 12.f});
}

void Game::handleInput(const sf::Event& event)
{
        if (event.type == sf::Event::KeyPressed)
        {
                switch (event.key.code)
                {
                        case sf::Keyboard::Escape:
                        {
                                util::Subject::send({ util::Message::PopState{} });
                                break;
                        }

                        default: break;
                }
        }
}

void Game::update()
{

}

void Game::draw(sf::RenderTarget& target)
{
        target.draw(this->currentLevel);
}

}

File: data/scripts/GameState/Game.lua
Game = {
    drawables = {
        {
            type = "rect tile map",
            tiles = {},
            tileSize = 16,
            tileIconSize = 16,
            texture = "tileset",
            origin = "center",
            position = {
                x = Settings.video.resolution.x / 2,
                y = Settings.video.resolution.y / 2,
            },
        }
    }
}

function Game:new(o)
    o = o or {}
    setmetatable(o, self)
    self.__index = self
    self.drawables[1].tiles = generate_dungeon {x = 37, y = 21}
    return o
end

function Game:handle_input(event)
    if event.type == "KeyPressed" then
        local kc = event.code
        if kc == Keyboard.Escape then
            pop_state()
        end
    end
end

function Game:update(dt)
end

function Game:draw()
    return self.drawables
end